### ISSUE TITLE
Feature request: fast copy between snapshot and live filesystem

### ISSUE THREAD

**Issue by shodanshok** [openzfs/zfs#8629]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| CentOS
Distribution Version	| 7.6
Linux Kernel	| 3.10.0-957.10.1.el7.x86_64
Architecture	| x86_64
ZFS Version	| 0.7.13-1
SPL Version	| 0.7.13-1

### Describe the problem you're observing
Depending on file size, restoring a file from a snapshot can be very slow. Moreover, it will heavily impact both I/O load and MRU cache. Finally, it will consume as much additional space as the size of the file to be restored.

*A fast copy operation will not only be faster but, more importantly, will not impose heavy I/O load nor additional space consumption.*

Please note that the usual suggestions (granular dataset, snapshot rollback and clones) are not a silver bullet:

- at the extreme, granular dataset means a new dataset for each big file. This is difficult to implement and follow regularly;

- zfs snapshot rollback are destructive: you can not rollback to a previous version while preserving the current one;

- clones can not be unlinked from their parent snapshots (and `promote` has its limitation also), meaning that significant space is bound to be lost in the long run.

### Describe how to reproduce the problem
```
# a 1 GB file was created under tank/test and snap1 was taken
[root@singularity ~]# ls -alh /tank/test/test.img 
-rw-r--r--. 1 root root 1.0G Apr 15 10:17 /tank/test/test.img

# 1 GB was used by the dataset, 0 B by the snapshot
[root@singularity ~]# zfs list tank/test -r -t all
NAME              USED  AVAIL  REFER  MOUNTPOINT
tank/test        1.00G  1.65T  1.00G  /tank/test
tank/test@snap1     0B      -  1.00G  -

# drop cache and recover the file; note the slow recovery rate
[root@singularity ~]# echo 3 > /proc/sys/vm/drop_caches 
[root@singularity ~]# dd if=/tank/test/.zfs/snapshot/snap1/test.img of=/tank/test/test.img bs=1M
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB) copied, 15.6934 s, 68.4 MB/s

# 2 GB are now used by the dataset, with 1 GB allocated by the snapshot
[root@singularity ~]# zfs list tank/test -r -t all
NAME              USED  AVAIL  REFER  MOUNTPOINT
tank/test        2.00G  1.65T  1.00G  /tank/test
tank/test@snap1  1.00G      -  1.00G  -
```

### Ideas for workarounds

**1. Enable hardlinking (`ln`) between a snapshot and the live filesystem.** As the snapshot is read-only, new writes to the hardlink will not overwrite any previous data.

**2. Create a dedicated command (ie: `zfs copy`) to instruct zfs to create a new file with the same block mappings.** `zdb -ddddd tank/test@snap1 2`shows the following output:
```
...
Indirect blocks:
               0 L2   1:3c0909b000:1000 20000L/1000P F=8192 B=89457/89457
               0  L1  0:3e07c19000:b000 20000L/b000P F=1024 B=89455/89455
               0   L0 0:3e05913000:20000 20000L/20000P F=1 B=89454/89454
           20000   L0 1:3c02dfe000:20000 20000L/20000P F=1 B=89454/89454
           40000   L0 1:3c03f3e000:20000 20000L/20000P F=1 B=89454/89454
           60000   L0 1:3c03f69000:20000 20000L/20000P F=1 B=89454/89454
           80000   L0 1:3c05049000:20000 20000L/20000P F=1 B=89454/89454
           a0000   L0 1:3c06af6000:20000 20000L/20000P F=1 B=89454/89454
           c0000   L0 1:3c05e81000:20000 20000L/20000P F=1 B=89454/89454
           e0000   L0 1:3c064b2000:20000 20000L/20000P F=1 B=89454/89454
          100000   L0 0:3e06a92000:20000 20000L/20000P F=1 B=89454/89454
          120000   L0 0:3e06ade000:20000 20000L/20000P F=1 B=89454/89454
          140000   L0 1:3c07c6f000:20000 20000L/20000P F=1 B=89454/89454
          160000   L0 0:3e07bd9000:20000 20000L/20000P F=1 B=89454/89454
          180000   L0 0:3e08c00000:20000 20000L/20000P F=1 B=89454/89454
...
```
If we can preserve the original mapping when creating a new file, it should be possible to "fast copy" any data from a previous snapshot.

**Comment by scineram** [comment]
There could be any number of snapshots inbetween also missing the file. So these could only work if the restoration is from the most recent snapshot.

**Comment by ptx0** [comment]
use nopwrite..

**Comment by jwittlincohen** [comment]
For those that don't know, this is what nop-write refers to:

"ZFS supports end-to-end checksumming of every data block. When a cryptographically secure checksum is being used (and compression is enabled) OpenZFS will compare the checksums of incoming writes to checksum of the existing on-disk data and avoid issuing any write i/o for data that has not changed. This can help performance and snapshot space usage in situations were the same files are regularly overwritten with almost-identical data (e.g. regular full-backups of large random-access files)."

To use nop-write, you need to use a cryptographically secure checksum (sha256, sha512, skein, or edonr), compression must be enabled, and the tunable `zfs_nopwrite_enabled` must be enabled.  edonr is the fastest of the secure checksums (350% faster than SHA256).

**Comment by shodanshok** [comment]
@ptx0 nop writes will avoid wasting space, but high I/O load remains (you need to read the entire source file). Moreover, it does not solve the "destructive rollback" issue. Don't get me wrong: nop writes *are* useful, just they don't solve the raised issues. I seem to remember that @behlendorf considered something similar in the past; am I right?

**Comment by dweeezil** [comment]
This is essentially a request for reflinks.  See #405.  I think that livelists, introduced by fast clone deletion, or at least a similar concept ought to make this reasonably possible.

**Comment by shodanshok** [comment]
@dweeezil Sure, reflinks would be a natural approach to provide such feature. Anyway I think it is so useful that, irrespective of the specific implementation, it would be very handy to have in zfs.

**Comment by amotin** [comment]
@voronind-com I have vague memories that `cp` with `--reflink[=always]` uses `FICLONE`, which thanks to Linux kernel can't clone between file systems.  Have you tried `--reflink=auto`, which IIRC was using `copy_file_range()`, which is more forgiving?

**Comment by voronind-com** [comment]
> [@voronind-com](https://github.com/voronind-com) I have vague memories that `cp` with `--reflink[=always]` uses `FICLONE`, which thanks to Linux kernel can't clone between file systems. Have you tried `--reflink=auto`, which IIRC was using `copy_file_range()`, which is more forgiving?

Thank you for your answer!

I played around with strace and results are sad. GNU coreutils try to use `ficlone`, but fail as Linux forbids it across different mountpoints. Then they fallback to full block copies. I could not make GNU coreutils to use `copy_file_range` in any way.
But! To my surprise, rust's `std::fs::copy` uses `copy_file_range` when `ficlone` is not possible. So I installed `uutils-coreutils` and now it works flawlessly by default with no parameters. I tested cp-ing 1 TB from a snapshot only took about 10 seconds and no increase in ALLOC, only in REFER.

Hope this helps other folks in the future. Also hope that either coreutils would fallback to `copy_file_range` or Linux would lift the `ficlone` restriction.
